js/server/modules/@arangodb/actions.js

```diff
@@ -1156,13 +1154,38 @@ function firstRouting (type, parts, routes, rawParts) {
 
 function routeRequest (req, res, routes) {
   if (routes === undefined) {
-    var dbname = arangodb.db._name();
+    let internalRoute = req.url.match(/^\/_(admin|api)\//);
+
+    const dbname = arangodb.db._name();
+    if (internalRoute) {
+      // internal route, i.e. /_admin, /_api, etc.
+      if (!InternalRoutingList[dbname]) {
+        // Do not join these two lines. If so there is a chance
+        // the RoutingList will contain undefined afterwards.
+        // i have not found out why this is the case.
+        const r = buildRouting(dbname, true);
+        InternalRoutingList[dbname] = r;
+      }
+      routes = InternalRoutingList[dbname];
+    } else {
+      // user-defined, i.e. custom Foxx route
+      if (!CustomRoutingList[dbname]) {
+        if (global.KEYSPACE_EXISTS('FoxxFirstSelfHeal')) {
+          // still waiting for initial selfHeal to execute
+          throw new internal.ArangoError({
+            errorNum: internal.errors.ERROR_HTTP_SERVICE_UNAVAILABLE.code,
+            errorMessage: "waiting for initialization of Foxx services in this database"
+          });
```

Comment:
I'm not sure if this would work better if it used `resultException` directly (see line 1937, but followed by a `return` to bail out). Unhandled exceptions like this are intercepted in `js/actions/api-system.js` line 53 and wrapped in a generic 500, which is why we're seeing 500 instead of 503.